transactor/src/main/scala/io/mediachain/transactor/Copycat.scala

```diff
@@ -44,32 +48,204 @@ object Copycat {
       server
     }
   }
+
+  class ClientException(what: String) extends RuntimeException(what)
+
+  sealed abstract class ClientState
+  
+  object ClientState {
+    case object Connected extends ClientState
+    case object Suspended extends ClientState
+    case object Disconnected extends ClientState
+  }
+
+  trait ClientStateListener {
+    def onStateChange(state: ClientState): Unit
+  }
   
   class Client(client: CopycatClient) extends JournalClient {
+    import scala.concurrent.ExecutionContext.Implicits.global
+    import ClientState._
+    
+    private var shutdown = false
+    private var state: ClientState = Disconnected
+    private var server: Option[String] = None
+    private var reconnectThread: Option[Thread] = None
     private var listeners: Set[JournalListener] = Set()
+    private var stateListeners: Set[ClientStateListener] = Set()
+    private val logger = LoggerFactory.getLogger(classOf[Client])
+    private val random = new Random
+    private val timer = new Timer(true) // runAsDaemon
+ 
+    client.onStateChange(new Consumer[CopycatClient.State] {
+                              def accept(state: CopycatClient.State) {
+                                onStateChange(state)
+                              }
+    })
+    
+    // submit a state machine operation with retry logic to account
+    // for potentially transient client connectivity errors
+    val maxRetries = 5
+    private def submit[T](op: Operation[T], retry: Int = 0): Future[T] = {
+      (state, shutdown) match {
+        case (_, true) =>
+          Future {throw new ClientException("Copycat client has shutdown")}
+
+        case (Disconnected, _) => 
+          Future {throw new ClientException("Copycat client is disconnected")}
+
+        case (Suspended, _) =>
+          logger.info("Copycat client is suspended; delaying "+ op)
+          backoff(1).flatMap { _ =>
+            logger.info("Retrying operation " + op)
+            submit(op, retry)
+          }
+          
+        case (Connected, _) =>
+          val fut = FutureConverters.toScala(client.submit(op))
```

Comment:
it is interacted in an evented context with submits originating in outside threads and entering the event loop through the copycat completable futures.
I don't think we need to synchronize any variables, the races with shuting down are resolved with the backoff retries.